src/gluonts/torch/model/deepar/_network.py

```diff
@@ -0,0 +1,450 @@
+            )
+        )
+
+        future_samples = []
+
+        # for each future time-units we draw new samples for this time-unit and update the state
+        for k in range(self.prediction_length):
+            # (batch_size * num_samples, 1, *target_shape, num_lags)
+            lags = self.get_lagged_subsequences(
+                sequence=repeated_past_target,
+                indices=self.shifted_lags,
+                subsequences_length=1,
+            )
+
+            # (batch_size * num_samples, 1, *target_shape, num_lags)
+            lags_scaled = lags / repeated_scale.unsqueeze(-1)
+
+            # from (batch_size * num_samples, 1, *target_shape, num_lags)
+            # to (batch_size * num_samples, 1, prod(target_shape) * num_lags)
+            input_lags = lags_scaled.reshape(
+                (-1, 1, prod(self.target_shape) * len(self.lags_seq))
+            )
+
+            # (batch_size * num_samples, 1, prod(target_shape) * num_lags + num_time_features + num_static_features)
+            decoder_input = torch.cat(
+                (
+                    input_lags,
+                    repeated_time_feat[:, k : k + 1, :],
+                    repeated_static_feat,
+                ),
+                dim=-1,
+            )
+
+            # output shape: (batch_size * num_samples, 1, num_cells)
+            # state shape: (batch_size * num_samples, num_cells)
+            rnn_outputs, repeated_states = self.rnn(
+                decoder_input, repeated_states
+            )
+
+            distr_args = self.proj_distr_args(rnn_outputs)
+
+            # compute likelihood of target given the predicted parameters
+            distr = self.distr_output.distribution(
+                distr_args, scale=repeated_scale
+            )
+
+            # (batch_size * num_samples, 1, *target_shape)
+            new_samples = distr.sample()
+
+            # (batch_size * num_samples, seq_len, *target_shape)
+            repeated_past_target = torch.cat(
+                (repeated_past_target, new_samples), dim=1
+            )
+            future_samples.append(new_samples)
+
+        # (batch_size * num_samples, prediction_length, *target_shape)
+        samples = torch.cat(future_samples, dim=1)
+
+        # (batch_size, num_samples, prediction_length, *target_shape)
+        return samples.reshape(
+            (
+                (-1, self.num_parallel_samples)
+                + (self.prediction_length,)
+                + self.target_shape
+            )
+        )
+
+    def forward(
+        self,
+        feat_static_cat: torch.Tensor,
+        feat_static_real: torch.Tensor,
+        past_time_feat: torch.Tensor,
+        past_target: torch.Tensor,
+        past_observed_values: torch.Tensor,
+        future_time_feat: torch.Tensor,
+        future_target: Optional[torch.Tensor] = None,
+        sampling: Optional[bool] = None,
+    ) -> Union[torch.Tensor, torch.distributions.Distribution]:
```

Comment:
(1) `Union` types don't work with TorchScript
(2) `torch.distributions.Distribution` cannot be moved between devices (they are missing a `to` method). Not sure if training on GPUs is possible when doing it like this.

Reply:
I would rather always return a `torch.Tensor` and construct the distribution object in the LightningModule

Reply:
Good points. It is hard however to make this type-stable if we want to use it for both training and prediction: to return a distribution you would need to return a tuple of tensors.